### PR TITLE
Improvements on changelog to be able to deliver reports

### PR DESCRIPTION
**Pull request by jmelis**
This PR aims to fix several issues that were present:

* Refactor to separate rendering from rest of code
* Services are grouped by url and they're only checked out once
* Add multiple output formats
* Fix problem with services that have 'none' as hash
* Improve efficiency by eliminating redundant actions
* General readability issues
* Fetch repository url to be displayed in report

### CODE REVIEW

File: saasherder/changelog.py [hunk 1]
```diff
@@ -1,140 +1,186 @@
...
+def run(cmd):
+    logger.info("$ {}".format(cmd))
+    p = Popen(cmd, stdin=PIPE, stdout=PIPE, stderr=PIPE, shell=True)
+    stdout, stderr = p.communicate()
+    if p.returncode != 0:
+        raise Exception("$ {} FAILED with exit code {}"
+                        .format(cmd, p.returncode))
 
-class Changelog(object):
+    return stdout.decode('utf-8')
```
Comment (jaseemabid) [inline saasherder/changelog.py:21]: +1. This looks neater than the static method.

File: saasherder/changelog.py [hunk 2]
```diff
@@ -1,140 +1,186 @@
...
 
-    def __init__(self, parent):
-        self.parent = parent
+    def plain(self):
+        header_template = "Context changes: {old:.8}..{new}\n{url}/compare/{old:.8}...{new}\n"
 
-    @staticmethod
-    def diff(now, previous):
-        """Return the difference between 2 service lists
+        section_template = textwrap.dedent("""\
```
Comment (jaseemabid) [inline saasherder/changelog.py:33]: TIL textwrap. +1

File: saasherder/changelog.py [hunk 3]
```diff
@@ -1,140 +1,186 @@
...
-    def last_changed(self, service, commit):
-        """Get the commit time for service at specific commit"""
+    def __init__(self, parent):
+        self.parent = parent
 
-        worktree = self.worktree(service['name'])
-        cmd = "cd {} && git log -1 --format=%cd {}".format(worktree, commit)
+        # Key: url of the service
+        # Value: list of services that have changed
+        self.changed_services = []
```
Comment (jaseemabid) [inline saasherder/changelog.py]: You are using an array and a hash with the same name. Could you simplify this a bit maybe?

File: saasherder/changelog.py [hunk 4]
```diff
@@ -1,140 +1,186 @@
...
+                else:
+                    diff_item = dict(name=service_name,
+                                     names=[service_name],
+                                     url=url,
+                                     new=service_hash,
+                                     old=prev_service_hash)
+
+                    changed_services[url] = diff_item
+
+        self.changed_services = changed_services.values()
```
Comment (jaseemabid) [inline saasherder/changelog.py]: I'd avoid this line if I can - mutating variables in a different scope.

File: saasherder/changelog.py [hunk 5]
```diff
@@ -148,40 +194,54 @@ def generate(self, context, old, new):
...
+            branch = run("git rev-parse --short HEAD").strip()
             logger.info("HEAD at {} before generating changelog"
                         .format(branch))
 
+        # Resolve commit if a date was passed in 'old'
+        try:
+            dt_old = parse(old)
+            old = run("git rev-list -1 --before='{}' {}".format(dt_old, branch)).strip()
+        except:
+            pass
```
Comment (jaseemabid) [inline saasherder/changelog.py]: 1. Except a more specific class of Error here.
2. Should you `exit(1)` instead of pass?

**Review by jaseemabid — CHANGES_REQUESTED** [review]
Requesting minor changes, but the patch looks quite good! I like the small functions :+1: but would avoid the mutations if I could.

File: saasherder/changelog.py (hunk 5 above)
Comment (jmelis) [reply saasherder/changelog.py]: I believe this is correct, `old` and `new` can be one of three things:

- a branch, for example: master
- a commit (short or long), for example: c589a14
- a date, for example: 2018-02-21 10:00

This piece of code will try to detect if it's a date (by doing dateutils.parse). If that parsing fails it shouldn't do anything and keep the original value as it's a commit or a branch. However if the parsing succeeds, then we need to find the commit associated to that time.

However I will refactor this a bit and add some comments, if it wasn't clear when reading the code it means that I need to fix it :)

File: saasherder/changelog.py (hunk 1 above)
Comment (jmelis) [reply saasherder/changelog.py:21]: I was thinking of putting it in a separate module and doing a few more convenience methods. For instance, with this method we're are losing stderr...

File: saasherder/changelog.py (hunk 2 above)
Comment (jmelis) [reply saasherder/changelog.py:33]: I also TIL'd just a couple of days ago! it's great and it's standard library!

File: saasherder/changelog.py (hunk 5 above)
Comment (jmelis) [reply saasherder/changelog.py]: You are right about excepting the specific class of Error, I hadn't read that. Will do.

File: saasherder/changelog.py (hunk 3 above)
Comment (jmelis) [reply saasherder/changelog.py]: Agreed.

File: saasherder/changelog.py (hunk 4 above)
Comment (jmelis) [reply saasherder/changelog.py]: I'm unsure what you mean? Can you further explain?

**Review by jaseemabid — APPROVED** [review]
LGTM :100: :+1:

**Pull request merged**